core/src/main/java/io/onetable/client/TableFormatClientFactory.java

```diff
@@ -39,16 +38,22 @@ public static TableFormatClientFactory getInstance() {
   }
 
   public TargetClient createForFormat(
-      TableFormat tableFormat, PerTableConfig perTableConfig, Configuration configuration) {
-    switch (tableFormat) {
-      case ICEBERG:
-        return new IcebergClient(perTableConfig, configuration);
-      case DELTA:
-        return new DeltaClient(perTableConfig, configuration);
-      case HUDI:
-        return new HudiTargetClient(perTableConfig, configuration);
-      default:
-        throw new NotSupportedException("Target format is not yet supported: " + tableFormat);
+      TableFormat tableFormat, PerTableConfigImpl perTableConfig, Configuration configuration) {
+    TargetClient tc = createTargetClientForName(tableFormat.name());
+    if (tc == null) {
+      throw new NotSupportedException("Target format is not yet supported: " + tableFormat);
+    }
+    tc.init(perTableConfig, configuration);
+    return tc;
+  }
+
+  public TargetClient createTargetClientForName(String tableFormatName) {
+    ServiceLoader<TargetClient> loader = ServiceLoader.load(TargetClient.class);
```

Comment:
I have not used the ServiceLoader before. Does the loader just pick up any implementation of the interface on the path and call the default constructor?

Reply:
Yes, it is rather elegant in its simplicity but does require some refactoring to align with the simplicity.
We use it heavily in Apache Knox.